pkg/controller/bastion/actuator_delete.go

```
@@ -0,0 +1,142 @@
+
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	ctrlerror "github.com/gardener/gardener/pkg/controllerutils/reconciler"
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (a *actuator) Delete(ctx context.Context, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) error {
+	logger := a.logger.WithValues("bastion", client.ObjectKeyFromObject(bastion), "operation", "delete")
+
+	opt, err := DetermineOptions(bastion, cluster)
+	if err != nil {
+		return err
+	}
+
+	credentials, err := openstack.GetCredentials(ctx, a.client, opt.SecretReference, false)
+	if err != nil {
+		return fmt.Errorf("could not get Openstack credentials: %w", err)
+	}
+
+	openstackClientFactory, err := a.openstackClientFactory.NewFactory(credentials)
+	if err != nil {
+		return fmt.Errorf("could not create openstack client factory: %w", err)
+	}
+
+	err = removeBastionInstance(logger, openstackClientFactory, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove bastion instance: %w", err)
+	}
+
+	err = removePublicIPAddress(logger, openstackClientFactory, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove public ip address: %w", err)
+	}
+
+	deleted, err := isInstanceDeleted(openstackClientFactory, opt)
+	if err != nil {
+		return fmt.Errorf("failed to check for bastion instance: %w", err)
+	}
+
+	if !deleted {
+		return &ctrlerror.RequeueAfterError{
+			RequeueAfter: 10 * time.Second,
+			Cause:        errors.New("bastion instance is still deleting"),
+		}
+	}
+
+	err = removeSecurityGroup(openstackClientFactory, opt)
+	if err != nil {
+		return fmt.Errorf("failed to remove security group: %w", err)
+	}
+	return nil
+}
+
+func removeBastionInstance(logger logr.Logger, openstackClientFactory openstackclient.Factory, opt *Options) error {
+	instances, err := getBastionInstance(openstackClientFactory, opt.BastionInstanceName)
```

err is not handled properly.